presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -445,4 +399,229 @@ private static TypeSignatureParameter applyBoundVariables(TypeSignatureParameter
+            }
+            Optional<Type> commonSuperType = typeManager.getCommonSuperType(originalType, actualType);
+            if (!commonSuperType.isPresent()) {
+                return SolverReturnStatus.UNSOLVABLE;
+            }
+            if (!satisfiesConstraints(commonSuperType.get())) {
+                // This check must not be skipped even if commonSuperType is equal to originalType
+                return SolverReturnStatus.UNSOLVABLE;
+            }
+            if (commonSuperType.get().equals(originalType)) {
+                return SolverReturnStatus.UNCHANGED;
+            }
+            bindings.setTypeVariable(typeParameter, commonSuperType.get());
+            return SolverReturnStatus.CHANGED;
+        }
+
+        @Override
+        public boolean isSatisfied(BoundVariables.Builder bindings)
+        {
+            return true;
+        }
+
+        private boolean satisfiesConstraints(Type type)
+        {
+            if (comparableRequired && !type.isComparable()) {
+                return false;
+            }
+            if (orderableRequired && !type.isOrderable()) {
+                return false;
+            }
+            if (requiredBaseName.isPresent() && !UNKNOWN.equals(actualType) && !requiredBaseName.get().equals(type.getTypeSignature().getBase())) {
+                // TODO: the case below should be properly handled:
+                // * `type` does not have the `requiredBaseName` but can be coerced to some type that has the `requiredBaseName`.
+                return false;
+            }
+            return true;
+        }
+    }
+
+    private class TypeWithLiteralParametersSolver
+            implements TypeConstraintSolver
+    {
+        private final TypeSignature formalTypeSignature;
+        private final Type actualType;
+
+        public TypeWithLiteralParametersSolver(TypeSignature formalTypeSignature, Type actualType)
+        {
+            this.formalTypeSignature = formalTypeSignature;
+            this.actualType = actualType;
+        }
+
+        @Override
+        public SolverReturnStatus update(BoundVariables.Builder bindings)
+        {
+            ImmutableList.Builder<TypeSignatureParameter> originalTypeTypeParametersBuilder = ImmutableList.builder();
+            List<TypeSignatureParameter> parameters = formalTypeSignature.getParameters();
+            for (int i = 0; i < parameters.size(); i++) {
+                TypeSignatureParameter typeSignatureParameter = parameters.get(i);
+                if (typeSignatureParameter.getKind() == ParameterKind.VARIABLE) {
+                    Long typeParameter = bindings.getLongVariable(typeSignatureParameter.getVariable());
+                    if (typeParameter == null) {
+                        // if an existing value doesn't exist for the given variable name, use the value that comes from the actual type.
+                        Optional<Type> type = typeManager.coerceTypeBase(actualType, formalTypeSignature.getBase());
```

Comment:
I think long parameters from `coerceTypeBase` should be used exclusively (e.g not mixed with existing bindings or provided long parameters). For example:
`function(decimal(x, 10))`
and you call it: `function(BIGINT 10)`. Then `x` would be bound `19` and the function signature would be `function(decimal(19,10))`. This is incorrect.

Maybe make such arguments like: `decimal(x,10)` illegal?